### ISSUE TITLE
ENH: diagnostic, outlier-influence, multicollinearity for penalized estimation

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#6331]
(I never looked at this.)

Is there anything special that we can add for penalized estimators or can we apply the standard methods?

current case: multicollinearity measures and vif for ridge regression. Does ridge regression or penalization "cure" multicollinearity?
some references in https://github.com/statsmodels/statsmodels/issues/6330#issuecomment-565079152

**Comment by danielchavez97** [comment]
From the paper you have linked in  [#6330 (comment)](https://github.com/statsmodels/statsmodels/issues/6330#issuecomment-565079152) it seems as if standard VIF measures are not appropriate for ridge regression.
It also seems like the authors have provided a closed form expression for calculating VIF in ridge contexts [here](http://r-es.org/9jornadasR/pdf/9JUR_paper_37.pdf)

While in practice I am sure many users would optimize their ridge regression to obtain an appropriate regularization parameter; in theory, a user might not wish to optimize this parameter or it may be too computationally expensive, in which case they may choose a parameter which would not 'cure' or 'alleviate' in any way the issues caused by multicollinearity- the most extreme case being if someone ran ridge with a regularization parameter very close to zero (essentially running OLS).

**Comment by josef-pkt** [comment]
I'm skeptical about Garcia et al claims, e.g. "The VIFs presented by Marquardt [3] and McDonald [4] for the RR are incorrectly calculated"

My impression about Garcia et al in several papers, that they are very picky about the definition of vif, and everything that doesn't correspond to their definition is "incorrect".

I have the old definition of vif for Ridge included in #2380.
If there are problems, and it seems like there are, then we will have to add options to compute various definitions of analogues to vif for the penalized case, and similarly for nonlinear cases like GLM.

(Aside: In general I'm not a fan of dropping variables based on vif or other automatic choices, but I added it in #2380 because it seems to be popular in machine learning or data science.)

as a diagnostic vif or similar multicollinearity measures after penalized estimation could be useful, e.g. based on the literature LASSO doesn't handle multicollinearity well, and elastic net is better in that case, but we still need to choose the weight between L1 and L2 penalization.

**Comment by josef-pkt** [comment]
aside: penalized estimation is not scale invariant or equivariant, in contrast to OLS. So vif based on standardized data would need to adjust for change in exog scales in the penalization (if we don't use standardized data already).

**Comment by josef-pkt** [comment]
a guess, not verified yet
the "new" definition of vif for Ridge is likely just `diag(inv(x'x + k I))`, which uses only the outer part of the Ridge sandwich cov_params. (that might be in other references)

`inv(x'x + k I) dot x'x` is a measure for the effect of penalization (this shows up also in several places)

**Comment by danielchavez97** [comment]
I am currently reading the paper which the car package uses (Fox, J. and Monette, G. (1992). ) to get a better sense of why that was chosen; otherwise, I think using `diag(inv(x'x + k I))` is the most straightforward way to apply VIF to ridge if we assume it requires no special treatment.

What I am realizing; however, is that taking this straightforward adaptation of what VIF is in OLS and applying it to Ridge, may provide information which may not be helpful to a user.

What I mean by this is that in a typical OLS use case, VIF is simply a function of the data, i.e the diagonal entries of the precision matrix of the data. On the other hand in the case of ridge, the penalization enters the picture and ends up determining the VIF in a way in which the VIF will be decreasing in the penalization parameter. So what would this really tell a user? That they can simply decrease the multicollinearity issue by increasing penalization? To me this simply seems like a restatement of the whole point one would use ridge regression in the first place no?

**Comment by danielchavez97** [comment]
Also I guess this ultimately gets into the question of what the use case is. Typically is important to those who wish to obtain consistent estimates of their regression coefficients, on the other hand, someone who simply cares about the predictive power of their model could probably care less about these estimates.

However in the case of the user interested in accurate parameter estimates, ridge would probably be used as a model/variable selection method (I am thinking of an econometric use case)

**Comment by josef-pkt** [comment]
> That they can simply decrease the multicollinearity issue by increasing penalization?

Yes.
I think the point in this issue is to get a measure for how much multicollinearity is left after penalization and to see the magnitude of the effect of penalization on multicollinearity and other diagnostic statistics.

That is, it can be useful to check vif in several stages of model building, e.g. vif on original data (centered or uncentered), and vif relevant for final cov_params estimate. 
In the case of WLS and GLM we would also have a weighted vif (i.e. using wexog), which might be the one for relevant for cov_params if there is no penalization.

vif or similar multicollinearity measures are also important for prediction, i.e. regularized estimator or variable selection will also improve prediction.

I would like to have similar diagnostic function across all main models, How useful they are will depend on the data and purpose of the model.

and a quote (I never used vif before starting to work on this for statsmodels)

```
After seeing that, I'm still puzzled why you'd even compute the VIFs. For y2, you have t statistics on the key variables of over 10. Why do you care about collinearity in such a case? Since I don't know what y1 and y2 are -- in fact, the round number of observations makes me think this is generated data -- I can't say more. xx and x1 aren't significant for explaining y1. It could be they're highly correlated with one another, but that isn't a problem for y2. Those standard errors properly capture all of the sampling uncertainty; there's nothing more to say, really.

I'll also note that the other explanatory variables with VIF > 10 are statistically significant. Again, why are you checking the VIFs?
```
https://www.statalist.org/forums/forum/general-stata-discussion/general/1522116-vif-test-of-multi-collinearity?p=1523468#post1523468

**Comment by josef-pkt** [comment]
related for pinv regularized singular exog

Marquardt, Donald W. 1970. “Generalized Inverses, Ridge Regression, Biased Linear Estimation, and Nonlinear Estimation.” Technometrics 12 (3): 591–612. https://doi.org/10.2307/1267205.

`diag(pinv(x'x))` except we use pinv(x) i.e. svd of x and not of x'x

however, this is regularized, i.e. singularity and infinite vif have been removed, see #2380

possible question: Do we have additional multicollinearity after removing singular vectors (small singular values) using pinv?
I never tried an example like that.

**Comment by josef-pkt** [comment]
I read Marquardt 1970

My impression is that his definition of vif combines the effect of penalization with multicollinearity, i.e. the measure is not variance inflation relative to an uncorrelated penalized estimate.
In this case variance inflation (deflation) can be smaller than 1, because ridge of pinv reduces the variance of the estimate, `cov_params`.

We still have the problem that Marquard's version and standard OLS vif are defined on correlation matrices, but penalized estimators are not invariant to rescaling, (unless we adjust the penalization weights as I did in GAM, but then we end up with generalized ridge and not simple k*eye ridge)

i.e. variance inflation/deflation of penalized estimation under multicollinearity and not just variance inflation for multicollinearity

**Comment by danielchavez97** [comment]
After having read these papers it seems as if there's no good answer as to whether ridge requires a special VIF equation, or whether a compelling one exists as of now. This leads me to believe that the current implementation in [#2380](https://github.com/statsmodels/statsmodels/pull/2380) should suffice as long as a user understands how the penalization can affect the VIF. 

Do you see any other issues that might arise from this implementation? 
I would be happy to look into anything else that may be of use/code up some implementations specified by the other papers and see how they compare.

**Comment by josef-pkt** [comment]
> Do you see any other issues that might arise from this implementation?

I don't have problems with this implementation.
The question is which information is provided by different versions of vif or similar measures, and is it useful to provide several different versions.

e.g. for Ridge we have `x'x`,  `x'x + k I`, product `(x'x + k I)^-1 x'x` and the current sandwich for from cov_params.
all 4 will give us different information about multicollinearity and/or amount of penalization.
The question is which are useful and can be explained to users.

I have not played with examples for this and don't have much intuition how informative those measures are.
(I'm still trying to figure out what options for scaling and constant handling we need for OLS vif and other collinearity measures.) 

More general to this issue is to look at this for other estimators than OLS, e.g. MLE in Logit, Poisson or GLM.
I think we might want something for the (penalized) score/estimating equation/moment condition, i.e. the inner part of a sandwich cov_params, or the nonlinear MLE equivalent of `x'x + k I` in the Ridge version.
(we might get already far using the WLS analogue where `x` is wexog instead of exog)

**Comment by danielchavez97** [comment]
I can definitely look into summarizing the differences between different measures and coming up with some examples which could go into documentation of the method. I can also look into this for the other estimators as well!

**Comment by josef-pkt** [comment]
I just remembered, that I'm using an option for cov_params
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/sandbox/regression/penalized.py#L180

I implemented a version of Generalized Ridge estimation, but based on Theil and Goldberger. In their interpretation the ridge penalty is prior information, i.e. information from a prior sample, which works similarly to a Bayesian prior.

In the ridge case this means that cov_params is scaled `x'x + k I` and not the sandwich form. This is also the default in GAM (GLM with L2 penalization) which follows the default in R mgcv.
GAM does not yet have a ridge sandwich version of cov_params, but `cov_type='HC0'` is similar but also robust to heteroscedasticity.

case of generalized inverse in OLS/WLS
Marquard 1970 uses `diag(pinv(x'x))`  for standardized x (x'x is correlation matrix) as the definition of vif for the regularized estimator.

**Comment by josef-pkt** [comment]
I did the derivation again for converting between generalized ridge with standardized/scaled and unscaled variables, i.e. how to transform the penalization

main parts

`A = x'x + a S`
define transform `x = D z`   or `z = D^{-1} x`    where D is full rank transformation, e.g. `diag(1 / std(x))`
(transpose below are missing if D is not symmetric)
then
`A = Dz'zD + a S = D (z'z + a D^{-1}  S D^{-1} ) D`

define transformed prior cov `B = D^{-1}  S D^{-1} ) D`

moment condition for generalized Ridge (assuming `E(y_prior) = 0`)
```
D (z'z + a B) D beta = x' y
(z'z + a B) gamma = D^{-1} x'y
(z'z + a B) gamma = z'y
```

where transformed params (beta coefficients) are `gamma = D beta`

for special case standard Ridge `S = eye`
we have 
`D = diag(std(x))`
`B = diag(1 / var(x))`

i.e. we want vif and similar to be based on `(z'z + a B)`

separate issue treatment of constant: is `x` in the above the 
- original exog, with a constant column and not demeaned or 
- demeaned without constant column

**Comment by josef-pkt** [comment]
outlier influence measures in GLM ridge regression
example Gamma regression

those just look like ridge adjustment to WLS versions of GLM irls 

Amin, Muhammad, Muhammad Amanullah, Muhammad Aslam, and Muhammad Qasim. 2019. “Influence Diagnostics in Gamma Ridge Regression Model.” Journal of Statistical Computation and Simulation 89 (3): 536–56. https://doi.org/10.1080/00949655.2018.1558226.
